### PR TITLE
[web_request] add handling of client_max_size in request body

### PR DESCRIPTION
**Pull request by pawelmhm**
## What do these changes do?

Adding size limits to request body so that malicious client can't cause trouble by trying to upload request with extremely big body. 

* default limit set to value same as in nginx 1mb;
* if client tries to upload body that exceeds this limit it will get 413 HTTP Error;
* limit can be updated by passing client_max_size argument to Application object

## Are there changes in behavior for the user?

Yes, too big uploads will raise HTTP 413 error.

## Related issue number

#1133 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] Add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [ ] Add a new entry to `CHANGES.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using `#issue_number` format at the end of changelog message. Use Pull Request number if there are no issues for PR or PR covers the issue only partially.

### CODE REVIEW

**Comment by pawelmhm** [comment]
I stumbled on two issues here:

1) there are `aiohttp.web_exceptions` and `aiohttp.errors`. Difference between the two is slightly unclear. They seem to contain similar code (e.g. 400, or 405 exception definittions) It would be nice to add some docstrings explaining their purposes. It seems like aiohttp.errors are mostly internal errors raised by library, and web_exceptions are intended to be raised from client code, is that correct? I added 413 to errors, is that ok? I cant add it to web_exceptions because of circular imports.

2) web_server.RequestHandler.handle_request will fail with UnboundLocalVariable error in finally block (line 69)  on every exception that is not HTTPException. It looks like a bug, am I right? `self._handler()` call in `try` part can raise multiple things that are not HTTPExceptions? It seems this was introduced recently, I did some change in code last weekend and behavior was different. Anyway I 'fixed' that, but without test, because I'm not exactly sure about this. Let me know if we should consider this bug or feature.

I guess code-cow failing because of missing test for 2.

File: aiohttp/web_reqrep.py [hunk 1]
```diff
@@ -70,6 +74,11 @@ def __init__(self, message, payload, transport, reader, writer,
         self._state = {}
         self._cache = {}
         self._task = task
+        if client_max_size:
+            max_size = client_max_size
+        else:
+            max_size = self._CLIENT_MAX_SIZE
```
Comment (asvetlov) [inline aiohttp/web_reqrep.py]: `max_size` could be `None` if user operates with low-level API (without `web.Application`, routers, signals, middlewares and other related stuffs).

File: aiohttp/web_reqrep.py [hunk 2]
```diff
@@ -363,6 +372,8 @@ def read(self):
             while True:
                 chunk = yield from self._payload.readany()
                 body.extend(chunk)
+                if len(body) >= self._client_max_size:
```
Comment (asvetlov) [inline aiohttp/web_reqrep.py]: `_client_max_size` could be `None`

File: aiohttp/web_reqrep.py [hunk 3]
```diff
@@ -363,6 +372,8 @@ def read(self):
...
+                    raise HttpRequestEntityTooLarge()
```
Comment (asvetlov) [inline aiohttp/web_reqrep.py]: For this particular case `from .web_exceptions import HTTPRequestEntityTooLarge` is allowed evil.
Yes, I know the import from function has ugly adore.
But I'm ok with this for the case.

File: aiohttp/web_server.py [hunk 4]
```diff
@@ -53,17 +54,21 @@ def handle_request(self, message, payload):
 
         request = self._request_factory(message, payload, self)
         self._request = request
+        resp = None
```
Comment (asvetlov) [inline aiohttp/web_server.py]: Yes, it's a bug. Thank you for finding.

File: aiohttp/web_server.py [hunk 5]
```diff
@@ -53,17 +54,21 @@ def handle_request(self, message, payload):
...
 
         try:
             try:
                 resp = yield from self._handler(request)
             except HTTPException as exc:
                 resp = exc
+            except HttpProcessingError as exc:
```
Comment (asvetlov) [inline aiohttp/web_server.py]: All exceptions not derived from `HttpProcessingError` should raise `500 Internal Server Errror` (now it's done by `server.py` code).

File: aiohttp/web_server.py [hunk 6]
```diff
@@ -53,17 +54,21 @@ def handle_request(self, message, payload):
...
+                raise exc
 
             resp_msg = yield from resp.prepare(request)
             yield from resp.write_eof()
         finally:
-            resp._task = None
+            if hasattr(resp, "_task"):
```
Comment (asvetlov) [inline aiohttp/web_server.py]: If you'll use `web_exceptions` this code is not needed.

**Comment by fafhrd91** [comment]
@pawelmhm are you still interested in working on this PR?

**Comment by AraHaan** [comment]
Uh @pawelmhm you might want to backup some of your changes and go with the latest commit on the master branch and work from there and remake this PR after that so that way it can fix the merge conflicts easier. Also it would be a good idea to also apply the requested changes that @asvetlov said.

**Comment by pawelmhm** [comment]
hey @AraHaan @fafhrd91 sorry for delay I was bit busy, yeah I'll do rebase against master and apply suggestions from @asvetlov I'll try to do this over the weekend

**Comment by fafhrd91** [comment]
Cool, thanks

File: aiohttp/web_reqrep.py [hunk 7]
```diff
@@ -372,6 +375,11 @@ def read(self):
             while True:
                 chunk = yield from self._payload.readany()
                 body.extend(chunk)
+                if self._client_max_size \
+                        and len(body) >= self._client_max_size:
+                    # local import to avoid circular imports
+                    from aiohttp import web_exceptions
```
Comment (fafhrd91) [inline aiohttp/web_reqrep.py:381]: Try to import `aiohttp.web`. All web exceptions are available i that namedspace

**Comment by fafhrd91** [comment]
thanks

File: aiohttp/web_reqrep.py (hunk 7 above)
Comment (pawelmhm) [reply aiohttp/web_reqrep.py:381]: ah I see you merged that, I can check if import of aiohttp.web works, should I open new PR with that + doc updates, adding myself to contributors?

Comment (fafhrd91) [reply aiohttp/web_reqrep.py:381]: I fixed import and added you to contributors already. thanks

**Pull request merged**